CBMC additional profiling info: Extend solver hardness to track clauses mapped to an instruction

natasha-jeppu opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
Currently solver hardness tracks number of clauses. This is an extension of solver hardness to track the clauses themselves. Provides a mapping between clauses and instructions.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Extend sat_hardnesst struct to keep track of the actual clauses, not just number of clauses

Currently, solver hardness tracks number of clauses and not the actual clauses for an
instruction. This mapping will help identfy code hotspots based on which of the clauses
belong to the SAT core.
This commit extends the sat_hardnesst structure to include mapping between clauses and
ssa expression.

Commit:
Add utility to collect clauses and display the clause set in json output for write-solver-stats-to

Clauses are stored as a sorted vector of integers.

Commit:
Add utility to collect clauses and display the clause set in json output for write-solver-stats-to

Clauses are stored as a sorted vector of integers.

clang format solver_hardness.cpp

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -91,10 +93,20 @@ void solver_hardnesst::register_clause(const bvt &bv)
 {
   current_hardness.clauses++;
   current_hardness.literals += bv.size();
+  std::vector<int> clause;
```
`int`? or do we want a type with a defined length like `int_64`?

peterschrammel replied:
❓ What are the memory and performance implications of storing a copy of the clauses?

natasha-jeppu replied:
I haven't explored this yet. Although a much more efficient approach would be to store some form of index (specifically an index interval) to some existing clause storage, but I am not sure if this can be done.

peterschrammel replied:
What are the metrics that you want to compute based on the clauses?

natasha-jeppu replied:
To identify which program lines contribute to the SAT core. So given this infrastructure we can map clauses from the SAT core directly to LOC.

peterschrammel replied:
What is a SAT core?

natasha-jeppu replied:
Referring to unsatisfiable core here.
https://en.wikipedia.org/wiki/Unsatisfiable_core

peterschrammel replied:
How are you going to compute the UNSAT core?

natasha-jeppu replied:
There are tools which compute the UNSAT core given a CNF formula and DRAT proof of unsatisfiability (these proofs have become a standard in the SAT competitions). CBMC is used to get the dimacs CNF formula (--dimacs) and the clause to LOC mapping with --write-solver-stats-to option.

peterschrammel replied:
If this can only be used wiith dimacs couldn't you just store the index (eg. line of the clause) in the dimacs output?

natasha-jeppu replied:
How do I access the dimacs clause index here? Also is the dimacs output ordered?
I will need to map lines of program code to indices in the dimacs output.

peterschrammel replied:
You can make `dimacs_cnft` (or probably `cnf_clause_listt`) implement `hardness_collectort`(in the same way other sat solvers implement that interface). It seems that the clauses are stored in order in `cnf_clause_listt`, so `dimacs_cnft` will print them in the same order as you get the `register_clause` calls. Of course, you'll have to test and confirm that.

natasha-jeppu replied:
I now keep track of the clause counter in `satcheck_minisat2_baset<T>::lcnf` that calls `register_clause` and store the counter value rather than the clause itself in solver hardness. Have added optional DEBUG statements to verify if these match corresponding lines in the dimacs output.

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -91,10 +93,20 @@ void solver_hardnesst::register_clause(const bvt &bv)
+    clause.push_back(signed_literal);
   }
+  clause.push_back(0);
+  std::sort(clause.begin(), clause.end());
```
Do we want to sort before or after we add the `0`?

natasha-jeppu replied:
Post processing of the SAT core cnf runs sort for the entire clause, including the zero. So might as well sort it here too.

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -91,10 +93,20 @@ void solver_hardnesst::register_clause(const bvt &bv)
+
+    int signed_literal = literal.var_no();
+    if(literal.sign())
+      signed_literal = -signed_literal;
```
Any chance this could overflow?

natasha-jeppu replied:
This is what is used inside CBMC already for the literal.dimacs() function. I'm assuming here that they have already taken into consideration any potential overflow issues. And also, they define the return type as int.

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -151,6 +163,16 @@ void solver_hardnesst::produce_report()
+      {
+        json_arrayt clause_json;
+        for(auto const &lit : clause)
+          clause_json.push_back(json_numbert{std::to_string(lit)});
```
`json_numbert` takes a string?

natasha-jeppu replied:
Yes, takes a string but the output is formatted json number.

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -26,6 +26,8 @@ operator+=(const solver_hardnesst::sat_hardnesst &other)
   clauses += other.clauses;
   literals += other.literals;
   variables.insert(other.variables.begin(), other.variables.end());
+  clause_set.insert(
```
Should we sort this?

natasha-jeppu replied:
I was worried about too many calls to sort().

danielsn commented on src/solvers/solver_hardness.cpp:
```diff
@@ -91,10 +93,20 @@ void solver_hardnesst::register_clause(const bvt &bv)
 {
   current_hardness.clauses++;
   current_hardness.literals += bv.size();
+  std::vector<int> clause;
```
Should we typedef this?

Commit:
Store clause counter value from satcheck_minisat instead of storing the clause itself

Commit:
Store clause counter value from satcheck_minisat instead of storing the clause itself

Add comment to explain solver_clause_num + 1 in solver hardness, change <int> to <size_t>

Remove sorting of clause set keeping performance in mind

Add additional utility to map program lines to clauses in --dimacs output

clang format

Commit:
Store clause counter value from satcheck_minisat instead of storing the clause itself

Add comment to explain solver_clause_num + 1 in solver hardness, change <int> to <size_t>

Remove sorting of clause set keeping performance in mind

Add additional utility to map program lines to clauses in --dimacs output

clang format

Commit:
Store clause counter value from satcheck_minisat instead of storing the clause itself

Add comment to explain solver_clause_num + 1 in solver hardness, change <int> to <size_t>

Commit:
Store clause counter value from satcheck_minisat instead of storing the clause itself

Add comment to explain solver_clause_num + 1 in solver hardness, change <int> to <size_t>

Remove sorting of clause set keeping performance in mind

natasha-jeppu commented on src/solvers/sat/satcheck_minisat2.cpp:
```diff
@@ -140,8 +140,10 @@ void satcheck_minisat2_baset<T>::lcnf(const bvt &bv)
-    with_solver_hardness(
-      [&bv](solver_hardnesst &hardness) { hardness.register_clause(bv); });
+    size_t solver_clause_num = clause_counter;
+    with_solver_hardness([&bv, &solver_clause_num](solver_hardnesst &hardness) {
+      hardness.register_clause(bv, solver_clause_num);
```
Pass the clause counter to hardness.register_clause(). Store the clause counter value rather than the clause themselves.

natasha-jeppu commented on src/solvers/solver_hardness.cpp:
```diff
@@ -89,24 +89,28 @@ void solver_hardnesst::register_assertion_ssas(
-  std::sort(clause.begin(), clause.end());
-  current_hardness.clause_set.push_back(clause);
+
+#ifdef DEBUG
+  std::cout << solver_clause_num << ": ";
+  for(const auto &literal : bv)
+    std::cout << literal.dimacs() << " ";
+  std::cout << "0\n";
+#endif
```
Debug statements to output 'clause_counter_value: clause' to cross check if they match the corresponding lines in dimacs output.

peterschrammel commented on src/solvers/sat/satcheck_minisat2.cpp:
```diff
@@ -140,8 +140,10 @@ void satcheck_minisat2_baset<T>::lcnf(const bvt &bv)
 
-    with_solver_hardness(
-      [&bv](solver_hardnesst &hardness) { hardness.register_clause(bv); });
+    size_t solver_clause_num = clause_counter;
```
Assignment seems redundant, just use `clause_counter` in line 145.

natasha-jeppu replied:
`clause_counter` is an inherited data member of `satcheck_minisat2_baset<T>` from `cnf_solvert` so I cannot pass it directly to the lambda capture list (not in scope). Will need to pass `this` instead. Thought it would be better to just pass a copy.

peterschrammel commented on src/solvers/solver_hardness.h:
```diff
@@ -50,6 +50,7 @@ struct solver_hardnesst
     size_t clauses = 0;
     size_t literals = 0;
     std::unordered_set<size_t> variables = {};
+    std::vector<int> clause_set = {};
```
```suggestion
    std::vector<size_t> clause_set = {};
```

natasha-jeppu replied:
Done.

peterschrammel commented on src/solvers/solver_hardness.cpp:
```diff
@@ -87,14 +89,28 @@ void solver_hardnesst::register_assertion_ssas(
+  std::cout << "0\n";
+#endif
+
+  current_hardness.clause_set.push_back(solver_clause_num + 1);
```
Please add a comment to explain why +1.

natasha-jeppu replied:
Done.

martin-cs commented on src/solvers/sat/satcheck_minisat2.cpp:
```diff
@@ -140,8 +140,22 @@ void satcheck_minisat2_baset<T>::lcnf(const bvt &bv)
 
     solver->addClause_(c);
 
+    // To map clauses to lines of program code, track clause indices in the
```
I don't want to be a downer on this but ...

1. There is not necessarily a 1-to-1 correlations between clauses and lines of code.  For example, rewriting and normalisation may mean that multiple lines reduce to the same shared term.  Which line should this be credited to?  This is worse when it is things like multiplication where additions from other operations can appear as sub-parts of the operator.

2. Number of variables and clauses is a *very* crude measure of SAT difficulty.  Reducing the number of clauses can make the problem harder; I have seen multiple instances of this.

martin-cs commented on src/solvers/solver_hardness.h:
```diff
@@ -105,7 +106,15 @@ struct solver_hardnesst
-  void register_clause(const bvt &bv);
+  /// \param cnf: processed clause
+  /// \param cnf_clause_index: index of clause in dimacs output
+  /// \param register_cnf: negation of boolean variable tracking if the clause
```
Please clarify this comment.

martin-cs commented on src/solvers/solver_hardness.h:
```diff
@@ -50,6 +50,7 @@ struct solver_hardnesst
     size_t clauses = 0;
     size_t literals = 0;
     std::unordered_set<size_t> variables = {};
+    std::vector<size_t> clause_set = {};
```
What is the performance cost of this?

martin-cs reviewed (commented):
I have done quite a bit of performance optimisation work on CBMC.  If you want a chat about it then drop me a line.
